man/man1/resend.1
.TH RESEND 1
.\" @(#) $Header: /xtel/pp/pp-beta/man/man1/RCS/resend.1,v 6.0 1991/12/18 20:43:41 jpo Rel $
.\"
.\" $Log: resend.1,v $
.\" Revision 6.0  1991/12/18  20:43:41  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
resend \- resubmit a message
.SH SYNOPSIS
.in +.5i
.ti -.5i
resend \%[\-u\ username] \%[\-t] \%[\-c] \%recipients...
.in -.5i
.SH DESCRIPTION
.I Resend
is a small utility that forwards a message to more recipients. It does
this by attaching extra fields to the original message of the form
Resent-To: and Resent-Cc:.
.PP
It has the following options:
.TP
\-t
Take the next arguments up to the next flag as recipients to place in
the Resent-To: field.
.TP
\-c
Take the next arguments up to the next flag as recipients to place in
the Resent-Cc: field.
.TP
\-u\ username
Pretend to this user when resubmitting. This is dependant on
authentication checks and non privelidged users will normally fail.
.SH FILES
None
.SH "SEE ALSO"
\fIPP Manual \- Volume 3: User Manual\fP
.SH "BUGS"
Not many.


CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Uip/resend/resend.c
/* resend.c: resend a message onwards */

# ifndef lint
static char Rcsid[] = "@(#)$Header: /xtel/pp/pp-beta/Uip/resend/RCS/resend.c,v 6.0 1991/12/18 20:42:09 jpo Rel $";
# endif

/*
 * $Header: /xtel/pp/pp-beta/Uip/resend/RCS/resend.c,v 6.0 1991/12/18 20:42:09 jpo Rel $
 *
 * $Log: resend.c,v $
 * Revision 6.0  1991/12/18  20:42:09  jpo
 * Release 6.0
 *
 */



#include "util.h"
#include "prm.h"
#include "q.h"
#include "adr.h"
#include "retcode.h"
#include <pwd.h>
#include <signal.h>
#include <sys/stat.h>
#include <varargs.h>

extern	char *strdup();
extern UTC	utclocalise();

extern char *loc_dom_site;
extern char *local_822_chan;
extern char *hdr_822_bp;
extern char	*ia5_bp;

static char   *username = (char *) 0;

#define ADR_TO	1
#define ADR_CC	2
static int	list = ADR_TO;

#define MAXADRS 100
static char	*toadrs[MAXADRS];
static int	tocnt = 0;
static char	*ccadrs[MAXADRS];
static int	cccnt = 0;

static RP_Buf	rps;
static RP_Buf	*rp = &rps;
static struct prm_vars prm;
static Q_struct qs;

static void adios ();

#define prefix(a,b)	(lexnequ ((a), (b), strlen(a)) == 0)

main (argc, argv)
int     argc;
char   *argv[];
{
	char	*cp;

	uip_init (argv[0]);

	pgminit ();

	if (argc <= 1)
		adios (NULLCP, "No addresses given");

	while (--argc > 0) {
		cp = *++argv;
		if (*cp == '-')
			switch (cp[1]) {
			    case 't':
				list = ADR_TO;
				break;
			    case 'c':
				list = ADR_CC;
				break;
			    case 'u':
				username = cp+2;
				if (!*username) { username = *++argv; argc--; }
				username = strdup(username);
				fprintf(stderr, "Username = `%s'\n", username);
				break;
			    default:
				adios (NULLCP, "Bad switch %s", cp);
			}
		else if (list == ADR_TO)
			toadrs[tocnt++] = cp;
		else	ccadrs[cccnt++] = cp;
	}
	fprintf(stderr, "Now send mail\n");
	sendmail ();
	exit (0);
}

pgminit ()
{
	extern struct passwd *getpwuid ();
	extern char *getmailid ();
	struct passwd  *pwdptr;
	int     realid;

	if (!username)
	{	realid = getuid ();

		if ((pwdptr = getpwuid (realid)) == (struct passwd *) NULL)
			err_abrt (RP_PARM, "Unable to locate user's name");

		username = strdup (pwdptr -> pw_name);
	}
}

sendmail ()
{
	ADDR	*adr;
	int i;

	prm_init (&prm);
	q_init (&qs);

	qs.inbound = list_rchan_new (loc_dom_site, local_822_chan);
	qs.encodedinfo.eit_types = list_bpt_new (hdr_822_bp);
	list_bpt_add (&qs.encodedinfo.eit_types, list_bpt_new (ia5_bp));

	if (rp_isbad (io_init (rp)) ||
	    rp_isbad (io_wprm (&prm, rp)) ||
	    rp_isbad (io_wrq (&qs, rp)))
		adios (NULLCP, "Unable to start submit: %s", rp -> rp_line);

	adr = adr_new (username, AD_822_TYPE, 0);
	adr -> ad_status = AD_STAT_DONE;
	adr -> ad_resp = NO;
	if (rp_isbad (io_wadr (adr, AD_ORIGINATOR, rp)))
		adios (NULLCP, "Bad sender %s: %s",
		       username, rp -> rp_line);
	adr_tfree (adr);

	for ( i = 0; i < tocnt; i++) {
		adr = adr_new (toadrs[i], AD_822_TYPE, 0);
		if (rp_isbad (io_wadr (adr, AD_RECIPIENT, rp)))
			adios (NULLCP, "Bad recipeint %s: %s",
			       toadrs[i], rp -> rp_line);
		adr_tfree (adr);
	}
	for ( i = 0; i < cccnt; i++) {
		adr = adr_new (ccadrs[i], AD_822_TYPE, 0);
		if (rp_isbad (io_wadr (adr, AD_RECIPIENT, rp)))
			adios (NULLCP, "Bad recipeint %s: %s",
			       ccadrs[i], rp -> rp_line);
		adr_tfree (adr);
	}

	if (rp_isbad (io_adend (rp)) ||
	    rp_isbad (io_tinit (rp)) ||
	    rp_isbad (io_tpart (hdr_822_bp, 0, rp)))
		adios (NULLCP, "Can't initialise for text submission: %s",
		       rp -> rp_line);
	
	dumpheader();
	doresent();

	dobody ();
}

dumpheader()
{
	char	line[LINESIZE];

	while (fgets (line, LINESIZE, stdin) != NULL) {
		if (line[0] == '\n')
			break;
		if (prefix ("resent-", line) ||
		    prefix ("received:", line) ||
		    prefix ("x400-received", line) ||
		    prefix ("via:", line))
			if (rp_isbad (io_tdata ("Old-", 4)))
				adios (NULLCP, "Data Copy failed");
		if (rp_isbad (io_tdata (line, strlen (line))))
			adios (NULLCP, "Data copy failed");
	}
}

doresent()
{
	char    datbuf[64];
	UTC	now, lut;
	now = utcnow();
	lut = utclocalise(now);
	UTC2rfc(lut, datbuf);		/* rfc822 format date */
	free ((char *) lut);
	sndhdr ("Resent-Date:  ", datbuf);
	sndhdr ("Resent-From:", username);
	doto ();
	(void) sprintf (datbuf, "1.%s", ia5_bp);
	if ( rp_isbad (io_tdend (rp)) || rp_isbad (io_tpart (datbuf, 0, rp)))
		adios (NULLCP, "Can't setup for body part: %s", rp -> rp_line);
}

sndhdr (name, contents)
char    name[],
	contents[];
{
	char    linebuf[LINESIZE];

	(void) sprintf (linebuf, "%-10s%s\n", name, contents);
	if (rp_isbad (io_tdata (linebuf, strlen (linebuf))))
		adios (NULLCP, "Data Copy failed");
}

doto ()
{
	register int    i;

	for (i = 0; i < tocnt; i++)
		sndhdr ("Resent-To:", toadrs[i]);
	for (i = 0; i < cccnt; i++)
		sndhdr ("Resent-Cc:", ccadrs[i]);
}

dobody ()
{
	char    buffer[BUFSIZ];
	register int    i;

	while (!feof (stdin) && !ferror (stdin) &&
	    (i = fread (buffer, sizeof (char), sizeof (buffer), stdin)) > 0)
		if (rp_isbad (io_tdata (buffer, i)))
			adios (NULLCP, "Problem writing body");

	if (ferror (stdin))
		adios (NULLCP, "Problem reading body");
	if (rp_isbad (io_tdend (rp)) || rp_isbad (io_tend (rp)))
		adios (NULLCP, "Error terminating: %s", rp -> rp_line);
}

#ifndef lint
static void    adios (va_alist)
va_dcl
{
    va_list ap;

    va_start (ap);

    _ll_log (pp_log_norm, LLOG_FATAL, ap);
    
    va_end (ap);

    _exit (1);
}
#else
/* VARARGS2 */

static void    adios (what, fmt)
char   *what,
       *fmt;
{
    adios (what, fmt);
}
#endif
